xgcm/transform.py

```diff
@@ -0,0 +1,230 @@
+                alpha = (theta_hat_max - theta_hat_min) / (theta_max - theta_min)
+                # now assign based on this weight
+                output[j] += alpha * phi[i]
+
+
+def interp_1d_conservative(phi, theta, target_theta_bins):
+    """
+    Accumulate extensive cell-centered quantity phi into new vertical coordinate
+    defined by scalar theta.
+
+    Parameters
+    ----------
+    phi : array_like
+        Array of shape (..., n) defining an extensive quanitity in a cell
+        bounded by two vertices.
+    theta : array_like
+        Array of shape (..., n+1) giving values of scalar theta  on the
+        cell vertices. Phi is assumed to vary linearly between vertices.
+    target_theta_bins : array_like
+        Array of shape (m) defining the bounds of bins in which to accumulate
+        phi.
+
+    Returns
+    -------
+    phi_accum : array_like
+        Array of shape (..., m-1) giving the values of phi partitioned into
+        specified theta bins.
+    """
+
+    assert phi.shape[-1] == (theta.shape[-1] - 1)
+    assert target_theta_bins.ndim == 1
+    # flip theta if needed
+    target_diff = np.diff(target_theta_bins)
+    if all(target_diff < 0):
+        flip_switch = True
+        target_theta_bins = np.flip(target_theta_bins)
+    elif all(target_diff > 0):
+        flip_switch = False
+    else:
+        raise ValueError("Target values are not monotonic")
+
+    theta_1 = theta[..., :-1]
+    theta_2 = theta[..., 1:]
+    theta_hat_1 = target_theta_bins[:-1]
+    theta_hat_2 = target_theta_bins[1:]
+
+    out = _interp_1d_conservative(phi, theta_1, theta_2, theta_hat_1, theta_hat_2)
+    if flip_switch:
+        out = np.flip(out)
+    return out
+
+
+"""Mid level functions (xarray)"""
+
+
+def input_handling(func):
+    """Decorator that handles input naming for interpolations."""
+
+    @functools.wraps(func)
+    def wrapper_input_handling(*args, **kwargs):
+
+        # unpack args
+        phi, theta, target_theta_levels, phi_dim, theta_dim, target_dim = args
+
+        # complain if the target values are not provided as xr.dataarray
+        if not isinstance(target_theta_levels, xr.DataArray):
+            raise ValueError("`target_theta_levels` should be passed as xr.DataArray")
+
+        # rename all input dims to unique names to avoid conflicts in xr.apply_ufunc
+
+        temp_dim = "temp_dim_target"
+        target_theta_levels = target_theta_levels.rename({target_dim: temp_dim})
+
+        # The phi_dim doesnt matter for the final product, so just rename to
+        # # something unique to avoid conflicts in apply_ufunc
+        temp_dim2 = "temp_unique"
+        phi = phi.rename({phi_dim: temp_dim2})
+
+        # Execute function with temporary names
+        args = (phi, theta, target_theta_levels, temp_dim2, theta_dim, temp_dim)
+        value = func(*args, **kwargs)
+
+        # rename back to original name
+        value = value.rename({temp_dim: target_dim})
+
```

Comment:
Currently the DataArray `value` never  have a name. It might be better if we automatically assigned a name when `phi` itself has a name. We could choose either
- `phi.name`
- `phi.name + suffix` where `suffix` is something like `_transformed` or `_interpolated`. This suffix could possibly be user customizable.

Reply:
ah good catch, I was just focusing on the actual coordinate names. 
Let me work that in...

Reply:
Just pushed a fix for this. grid/axis.transform now have `suffix='_transformed'` as kwargs